proposals/0002-gep.md

```diff
@@ -0,0 +1,207 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# TITLE: Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers.
+
+## Motivation
+
+[TBU, what problem we are trying to solve, focus on users, use cases, and all important context]
+
+The Unified API holds all the logic needed to create a request to a provider. This solution allows the clients to be easily
+envoked by a provider pool.
+
+
+### Requirements
+
+- R1: User should be able to specify specific models or select auto mode (selection handled by Glide)
+- R2: Should be a Hertz Client
+- R3: Should take a user request as input
+- R4: Handles all provider specific logic
+- R5: Easily maintained
+
+
+## Design
+
+
+- A Hertz client that handles an incoming request and returns a reponse from that provider.
+- Provider specific clients
+
+### Routes
+```yaml
+routes:
+    chat: api/v1/chat
+    complete: /api/v1/complete #this is also know as generate in some providers
+
+    # want to avoid specifying a provider pool as in the future we may
+    # want to combine strategies e.g User employs a fallback strategy where
+    # the fallback model is the one with the lowest latency
+```
+
+#### User Request Schema for Chat Route
+
+Below is an example user request to the Chat route. The other routes will look very similar, if not identical. The provider specific params will look different.
+
+
+Example Request Body in YAML for readability & comments
+
+```yaml
+mode: fallback # selected strategy, required(?)
+globalParams: # Selected model inherits these where applicable, optional
+  temperature: 0.7
+  max_tokens: 100
+models:
+  - provider: openai # Ideally this field is optional but for MVP it may need to be required I think
+    apiKey: api-key # required
```

Comment:
Such request schema:
- duplicates a lot of Glide configuration that otherwise is hidden behind provider pool setup: https://github.com/modelgateway/geps/pull/2/files#diff-0e4c03c5c73029151e17a7f008f0e7a86d45cc14244a5567769a3929db9337f2R112-R126
- exposes routing detail to the client that they should probably remain under the curtain.

I was looking at this differently:
- Glide is configured with provider pools needed for applications to work. A good deal of configs are taken out of the business logic services
- The provider pools have names that could be referenced on the application side when they access Glide API like `/v1/{pool-id}/chat/` , for example.
- Then, the client sends requests where it specifies one unified set of params (and have a way to override that for a specific provider in the pool referencing it by provider ID in the pool or provider type e.g. openai, anthropic (but you could have a few openai items configured in the pool)

This way we decouple as much LLM management/rounting/config as possible leaving out the business application as dumb/simple/focused on the end goal as possible.

Let's debate this, because I believe this is a crucial point. 

Also, I will think what the unified request schema looks like in that case.

Reply:
This conversation is great. Lets dig into this a bit. Here is a simple user journey to help answer some questions I have and understand your perspective better.

- I'm an LLM dev and need to stand up a RAG Q&A app. How do I deploy this as you see it? 
- When I deploy it all my LLM config is now in the gateway and I have a single endpoint such as /v1/{pool-id}/chat/ to send messages to, correct? I basically just send messages to the endpoint no provider param details.

Are these assumptions correct?

Reply:
Will try to answer (and please feel free to challenge/clarify this further)

> I'm an LLM dev and need to stand up a RAG Q&A app. How do I deploy this as you see it?

In a RAG case, I imagine the following workflow:
- The LLM dev indexes their data in a vectorDB
- The LLM dev crafts prompt, search for relevant context in the vectorDB and adds it to the prompt. Also, they experiment with params and prompt to find a set that works the best
- Up to this point, this was experiments. Now it's time to make that app more production ready. The LLM dev deploys/uses hosted vectorDB. Along with that, they deploy Glide with one `default` provider pool. The pool contains one model they experimented
- The Q&A app uses Glide client to work with LLM. It also maintains the conversational memory and sends it as soon as a conversation develops. The app uses `/v1/default/chat` for that, for example.
- To make it more resilient, the LLM dev adds two more LLMs in the `default` pool.
- After more experiments and testing, turned out that one of fallback models in the pool works badly with the prompt used for the main model. The LLM dev overrides the prompt for that model, so now each time the application sends request to Glide, it specifies two prompts: the main one (used by two models) and one more for one of the fallback models.

In case a new use case is found for LLM in the system, the default pool could be reused or a new one created specific to that usecase.

So that would be my understanding.

> When I deploy it all my LLM config is now in the gateway and I have a single endpoint such as /v1/{pool-id}/chat/ to send messages to, correct?

Most of it. We still could allow to override some params on the fly that have something to do with how the model functions but not operational configs like API key, base URL (in case of on-prem models), etc.

Getting back to the current request schema. It basically defines a provider pool but does it dynamically. So theoretically, the client could send completely different pool config on every request. If so, how would we match/correlate important state like load balancing stats? By client identity that sends the request? Okay, but if that pool config was different in this request to the prev ones, would we need to forget the prev state/stats and start from scratch or we need to try to match some part of state like provider health omitting load balancing state? 
Unless really really needed and useful, this level of flexibility seems to greatly complicate things.

Reply:
Alright this is super helpful. I see scenario where you have a lot of different provider pools. Not sure that is necessarily a bad thing, rather just a function of this design. I'm aligned on this direction now and will update the get